### ISSUE TITLE
request: reliable means to interrogate sparsity

### ISSUE THREAD

**Issue by bkmartinjr** [theislab/anndata#326]
It would be very helpful if the API had a standard way to query the anndata X sparsity, and sparse format, independent of the backed mode.

In other words, a safe / reliable / performant equivalent equivalent of the following on memory-backed adata:
* `scipy.sparse.issparse(adata.X)` -- bool indicates if is sparse
* `adata.X.get_format()` -- return scipy sparse format, eg, 'csr'

Is there a way to do this now?  Or a hack to work around the lack of API today?

One possible hack-around relies on the weak test for the format_str internal to spmatrix (these are overly wordy for clarity):

```
from scipy import sparse

def issparse(adata):
    if adata.isbacked:
        return hasattr(adata.X, 'format_str')
    else:
        return sparse.issparse(adata.X) 

def get_sparse_format(adata):
    if issparse(adata):
        if adata.isbacked:
            return getattr(adata.X, 'format_str', 'und')
        else:
            return adata.X.getformat()
    return False
```

**Comment by flying-sheep** [comment]
If you just want to test sparsity, and not if `X` is some other kind of array-like thing, you’d try `scipy.sparse.issparse(adata.X)` or `isinstance(anndata.X, scipy.sparse.spmatrix)`. For the sparse format, you’d use `scipy.sparse.isspmatrix_csr` or `isinstance(..., ...csr_matrix)`, right?

This should actually work for `BackedSparseMatrix`, as all its subclasses actually inherit from the `spmatrix` subclasses:

https://github.com/theislab/anndata/blob/4440b90ff3dff213b4c512478e21426cf210368d/anndata/_core/sparse_dataset.py#L117

https://github.com/theislab/anndata/blob/4440b90ff3dff213b4c512478e21426cf210368d/anndata/_core/sparse_dataset.py#L143

**Comment by bkmartinjr** [comment]
> If you just had plain scipy.sparse classes and wanted to test sparsity, you’d use scipy.sparse.issparse(adata.X) or isinstance(anndata.X, scipy.sparse.spmatrix), right? 

Yes, that would be perfect.

> For the sparse format, you’d use scipy.sparse.isspmatrix_csr or isinstance(..., ...csr_matrix), right?

I would probably use `spmatrix.getformat()`, but that is equivalent to scipy.sparse.isspmatrix_csr(), and both should work if `adata.X` is a subclass of the appropriate spmatrix.

That said, in 0.7.1, it doesn't seem to work.  The type of `adata.X` is a SparseMatrix, not a BackedSparseMatrix.

```
>>> adata = anndata.read_h5ad('pbmc3k-CSC-gz.h5ad', backed='r')
>>> adata
AnnData object with n_obs × n_vars = 2638 × 1838 backed at 'pbmc3k-CSC-gz.h5ad'
    obs: 'n_genes', 'percent_mito', 'n_counts', 'louvain'
    var: 'n_cells'
    uns: 'draw_graph', 'louvain', 'louvain_colors', 'neighbors', 'pca', 'rank_genes_groups'
    obsm: 'X_pca', 'X_tsne', 'X_umap', 'X_draw_graph_fr'
    varm: 'PCs'
>>> adata.X
<HDF5 sparse dataset: format 'csc', shape (2638, 1838), type '<f4'>
>>> type(adata.X)
<class 'anndata._core.sparse_dataset.SparseDataset'>
>>> scipy.sparse.isspmatrix(adata.X)
False
>>> scipy.sparse.isspmatrix_csc(adata.X)
False
>>> isinstance(adata.X, scipy.sparse.spmatrix)
False
>>> isinstance(adata.X, scipy.sparse.csc_matrix)
False
```

The `SparseMatrix` class has some intriguing parameters (eg, `format_str`), but I assume those are private?

**Comment by flying-sheep** [comment]
That’s very weird, `X` is supposed to be a matrix class, not a HDF5 dataset.

**Comment by bkmartinjr** [comment]
LMK if you need a more complete test case.

**Comment by ivirshup** [comment]
@bkmartinjr You should be able to get some idea from `type(adata.X)`. 

Do you have a use case where it's important to know whether it's dense or sparse, but it's not important to know whether it's in memory or on disk?

It might be nice to have this in the future, but I'd really like anndata to be more agnostic about the type of the array. That is, someone defines an array type and it works fine without us having to approve it. These trait detection methods should also be implemented in upstream packages (I've suggested this here: https://github.com/scipy/scipy/issues/11565).

For right now, I think you could get away with:

```python
if isinstance(X, (scipy.sparse.spmatrix, SparseDataset)):
    # It's sparse
elif isinstance(X, (np.ndarray, h5py.Dataset)):
    # It's dense
```

@flying-sheep if the AnnData is backed and not a view, X is a `h5py.Dataset` or our `SparseDataset`.

`SparseDataset` should get exported, and we should make its `to_backed()` method private.

**Comment by flying-sheep** [comment]
Why isn’t X a matrix-like class that produces matrices when sliced?

**Comment by bkmartinjr** [comment]
> You should be able to get some idea from type(adata.X)

True, but doing that is what got this entire conversation started.  AFAIK, there are no public types usable in that manner.  As a hack, it is definitely an alternative.  I had picked `format_str` instead of the type, as it seemed to be more broadly backward compatible.

> It might be nice to have this in the future, but I'd really like anndata to be more agnostic 

Fully agree with this, as long as you define the minimum interface that the array must support.   If that abstract interface is sufficient, all would work fine.   

I consider sparsity to be necessary in that interface, as it is very hard to write high performance code, for large matrices, without knowing something about how you should partition and stride through it.   It is more important, IMHO, than `isbacked`.

> For right now, I think you could get away with:

This is similar to the hack I was using before, which broke when the internal class structure changed. 

TL;DR - I like your "agnostic" approach, but what it needs to work IMHO is:
* a "guaranteed" interface for all array-ish things that can be stashed in X
* if there are modes of behavior for X (like sparse/dense), interfaces that let users introspect these modes

**Comment by pablo-gar** [comment]
Coming in because I've ran into the exact same issue. Specifically I'm wondering if this will be fixed

```
adata = anndata.read("file.h5ad", "r")
type(adata.X)
# anndata._core.sparse_dataset.SparseDataset

adata = anndata.read("file.h5ad")
type(adata.X)
# scipy.sparse.csr.csr_matrix
```

Why is the type of X `SparseDataset` when read in backed mode?

**Comment by flying-sheep** [comment]
With #219 (backed zarr), the incantation for this would get harder. what’s the plan here @ivirshup?

**Comment by ivirshup** [comment]
Would still like to not implement our own solution, especially one we export, since I don't think we can reliably cover all cases. I would ideally like scipy to make `issparse` single dispatch, but idk if that's going to happen.

Realistically I think the thing to do will be something like:

```python
SPARSE_TYPES = (scipy.sparse.sparray, scipy.sparse.spmatrix, ...)

isinstance(x, SPARSE_TYPES)
```
